articles/per_package_documentation.md

```diff
@@ -0,0 +1,115 @@
+
+## Requirements
+
+The "Primary requirements" are those that must be in place for the system to be functional and achieve its purpose.
+The "Secondary requirements" are important features and functionality that are not necessary to roll out the first stage of implementation, but should be carried out as soon as possible following the initial roll out.
+
+### Primary requirements
+
+**All package documentation must be available with the rest of the ROS 2 documentation under a single domain**
+
+Package documentation will be treated the same as the generic ROS 2 documentation.
+Its presence as part of the docs site should be made well-known from the site's overview as well as from any entry point into the docs site, and be intuitive from the organization and layout of the site.
+
+**Package documentation must have an easily navigable and intuitive URL structure**
+
+Every package's "home page" should be reachable with minimal effort by the scheme `docs.ros.org/p/<package_name>/`, which will redirect to default language and version, `docs.ros.org/<lang>/<distro>/p/<package_name>/`  
+
+**Package documentation must be maintainable in its repository without going through a third party**
+
+Maintainers and contributors will only have to work on their package's documentation within its repository.
+The details of building and hosting will not be a concern of package maintainers.
+
+Despite being hosted alongside the generic documentation, working on a package's documentation will not require any work on the repositories of the generic documentation (currently `ros2/ros2_documentation`) or the site repository (currently `ros-infrastructure/rosindex`).
+
+**The system must support C++ and Python API docs generation**
+
+The system will automatically extract API docs for these languages from the source code in package repositories and build the output on the docs site.
+
+**The system must allow package documentation to be versioned per ROS 2 distribution**
+
+The docs site and buildfarm will allow documentation for the latest version of a package corresponding to each ROS 2 distribution the docs site supports versioning for.
+The package repository can still maintain it's own docs for previous versions.
+
+While multiple package versions per distribution will not initially be supported, the URL structure will support the addition of this feature in the future (mentioned in the Secondary requirements below).
+
+**The buildfarm must automatically build package documentation**
+
+Changes to the documentation in a package repository shouldn't require the maintainer to manually trigger a build.
+The changes should be automatically built and uploaded to the site.
+
+**The system must automatically generate content for a package so it's listed on docs.ros.org even if the package maintainer does not explicitly set up anything in the package repository**
```

Comment:
This requirement conflicts with this line from the background:

> Package documentation from the repositories is built on the ROS 2 infrastructure and deployed to the ROS 2 documentation site in an automatic process with no input from the maintainer besides an opt-in `doc` block in the distribution.yaml

While there's no conflict about not setting up anything in the package repository itself, the requirement says the system is opt-out while the line from the background is explicit that it's opt-in.

If the intention is that *some* page will be generated for every package but documentation will not be generated unless the package opts in, then perhaps clarify this requirement and the background so that is clear.

Reply:
I’d like something to autogenerate given many packages would otherwise have nothing. In ROS1 if you asked it to setup docs in Bloom it would generate API docs even if there was no doxygen which was awesome. Something minimal like that and the readme would at least be something.

Reply:
This requirement is slightly different from the one in the background section, but you're right, it's not clear. The background section is talking about building any actual documentation like API docs and such which would be opt in. 

This point is talking about generating some generic content just for an index of packages on the package documentation homepage, that'll include the package name and maybe some other generic info like maintainer, activity, etc. That would be opt out